src/wp-includes/option.php

```diff
@@ -2887,3 +2889,48 @@ function filter_default_option( $default_value, $option, $passed_default ) {
 
 	return $registered[ $option ]['default'];
 }
+
+/**
+ * Checks if two database values are equal.
+ *
+ * @since 6.4.0
+ * @access private
+ *
+ * @param mixed $old_value The old value to compare.
+ * @param mixed $new_value The new value to compare.
+ * @return bool True if the values are equal, false otherwise.
+ */
+function _is_equal_database_value( $old_value, $new_value ) {
+	$values = array(
+		'old' => $old_value,
+		'new' => $new_value,
+	);
+
+	foreach ( $values as $_key => &$_value ) {
+		/*
+		 * Special handling for false-ish values.
+		 * Empty strings in the database should be seen as equivalent to false-ish cache values.
+		 * Cast scalars to a string so type discrepancies don't result in cache misses.
+		 */
+		if ( false === $_value ) {
+			$_value = '0';
+		} elseif ( 'old' === $_key && '' === $_value ) {
+			$_value = '0';
+		} elseif ( is_scalar( $_value ) ) {
+			$_value = (string) $_value;
+		}
```

Comment:
I think there are a few things in here that are problematic, causing some of the test failures at least:
* Not all "false-y" values are equal: I think specifically, we should differentiate between values that in the database will evaluate to `''` vs `'0'`.
* So for example, `false` should be considered as `''`, not `'0'`, and an old value of `''` shouldn't become `'0'`.

I'm not sure this fixes all test failures, but I would maybe give this alternative a try to see what happens:
```suggestion
		if ( null === $_value || is_scalar( $_value ) ) {
			$_value = (string) $_value;
		}
```

Reply:
Thanks @felixarntz this did the trick for BC tests but need to adjust the test for `test_is_equal_database_value`

Reply:
@mukeshpanchal27 Yeah that makes sense. Given those tests are new, it's okay to adjust them, but the existing tests passing as is is most important.